### PR TITLE
Start exec command agent upon container startup

### PR DESCRIPTION
**Pull request by angelcar**
### Summary

Add ExecCommandAgent manager, which can idempotently start and restart the ExecCommandAgent with exponential back off and max retries.
### Implementation details

Create a new manager struct in order to abstract ExecCommandAgent away from the engine. Now the engine is much simpler since it's not aware of what it takes to start/restart ExecCommandAgent. Modified engine's UTs accordingly 

### Testing

* Manually tested that the ExecCommandAgent is started in the container. 
* Manually stopped ExecCommandAgent to verify it's restarted automatically

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/execcmdagent/manager.go
Comment (sharanyad) [inline agent/engine/execcmdagent/manager.go:29]: ideally, we would want the retries to finish within the health checkup interval (which is 15 mins), so this might need to be changed accordingly.

File: agent/engine/execcmdagent/manager_linux.go
Comment (sharanyad) [inline agent/engine/execcmdagent/manager_linux.go:41]: can this and below check be done in the parent method to fail early?

File: agent/engine/execcmdagent/manager_linux.go
Comment (sharanyad) [inline agent/engine/execcmdagent/manager_linux.go:54]: if this is a transient docker error, then returning an error triggering a submit status change as agent stopped may not make sense (exec agent could still be running). feels like unless we know for sure agent is not running, we should not trigger submit status change. same for docker timeout errors.

File: agent/engine/execcmdagent/manager_linux.go
Comment (sharanyad) [inline agent/engine/execcmdagent/manager_linux.go:87]: shouldn't this return an error since we reached max retries to start agent?

File: agent/engine/execcmdagent/manager_linux.go
Comment (sharanyad) [inline agent/engine/execcmdagent/manager_linux.go:69]: why is this check a requirement? ie. when do we anticipate a process already running?

File: agent/engine/docker_task_engine.go
Comment (sharanyad) [inline agent/engine/docker_task_engine.go:302]: on restart failing, the exec metadata needs to be reset https://github.com/aws/amazon-ecs-agent/pull/2648/files#diff-b9aa12c3425c34d6f330f28407f779f6L305

File: agent/engine/docker_task_engine.go
Comment (sharanyad) [inline agent/engine/docker_task_engine.go:1260]: feels like this should be outside the startContainer method - ie. container moving to RUNNING is blocked by starting exec agent here, when in real it does not affect start container flow.

File: agent/engine/execcmdagent/manager.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager.go:29]: This timeout is the total timeout, not an individual timeout. I set it to 10 mins thinking about the 15 mins of the health check.

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:54]: I see your point. I still want to signal something is wrong in this case. Will think more about it.

File: agent/engine/execcmdagent/manager.go
Comment (sharanyad) [reply agent/engine/execcmdagent/manager.go:29]: my bad, I read it as 30 minutes 😅

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:87]: This is a weird aspect of`retry.RetryWithBackoffCtx`. if max retries is reached AND there is an error, then `startErr` will store it. The return is only to break the internal loop of `retry.RetryWithBackoffCtx`. I thought about refactoring it, but it's being used in too many places. 
Other than the weird readability, it works as you can see in the UTs.

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:69]: As per the design doc, there might be customers already running their own ssm agent in their containers. 
For these cases we don't want to start another ssm agent.

File: agent/engine/docker_task_engine.go
Comment (angelcar) [reply agent/engine/docker_task_engine.go:302]: Good catch, I will reset it inside `RestartIfStopped` because we still need the metadata in the following scenario:
* Docker exec inspect fails: in this case we could not verify if the agent is running, in the next health check we need the previous metadata to do another docker exec inspect, if we remove it, we would lose track of it and it might still be running.

File: agent/engine/docker_task_engine.go
Comment (angelcar) [reply agent/engine/docker_task_engine.go:1260]: This is another weird thing of the engine; actually this doesn't block the start. I've observed that if the `client.StartContainer` call above (line 1200) succeeds, then container will be reported as RUNNING, because then docker reports the start to the client in ecs agent, which is listening for docker events.

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:41]: I thought about it, and I do it for Start, can do it for this one too, but I still want to keep this checks inside the functions because the engine is not very well encapsulated and in the future someone might call these functions without first checking if exec is enabled.

Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:41]: Actually, this is implicitly done because the health check process skips non-exec enabled tasks: https://github.com/aws/amazon-ecs-agent/pull/2648/files#diff-b9aa12c3425c34d6f330f28407f779f6R281

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:87]: Will see if I can use another construct of the `retry` package, but it wasn't much different.

File: agent/api/task/task_linux.go
Comment (sparrc) [inline agent/api/task/task_linux.go:58]: what's the status of this TODO?

File: agent/engine/docker_task_engine.go
Comment (sparrc) [inline agent/engine/docker_task_engine.go:303]: This is an API that CP is in process of implementing?

File: agent/engine/execcmdagent/manager_linux.go
Comment (sparrc) [inline agent/engine/execcmdagent/manager_linux.go:50]: why are we checking if it's a DockerTimeoutError if we're just going to return `false, err` either way?

File: agent/engine/execcmdagent/manager_linux.go
Comment (sparrc) [inline agent/engine/execcmdagent/manager_linux.go:61]: is this the exit code of ExecCommandAgent process or of the container? Is there documentation or design around different exit codes from ExecCommandAgent?

Comment (sparrc) [reply agent/engine/execcmdagent/manager_linux.go:61]: If there isn't probably better to remove the TODO, as it implies that ExecCommandAgent has logic around it's exit codes

File: agent/engine/execcmdagent/manager_linux.go
Comment (sparrc) [inline agent/engine/execcmdagent/manager_linux.go:132]: same exit code comment as above

File: agent/api/task/task_linux.go
Comment (angelcar) [reply agent/api/task/task_linux.go:58]: To be defined, the ssm agent is still working on this, so I want to make sure I don't forget by leaving this TODO here

File: agent/engine/docker_task_engine.go
Comment (angelcar) [reply agent/engine/docker_task_engine.go:303]: Yes it is, Also, Ray is gonna work in integrating this part, leaving this TODOs so we have a reference of where to put things

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:50]: Yeah, we were contemplating to have different treatments, but I removed that

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:61]: I prefer to leave the TODO here, as there are discussions in progress about the exit codes. Before this feature branch is merged, all the `[ecs-exec]` will be addressed, either by removing them or by implementing additional stuff.

Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:61]: Oh, and the exit code is the exit code of ExecCommandAgent

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:132]: ditto

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:54]: Ended up adding retries for the inspect. If the inspect fails after all retries, then I return the error, I think it's reasonable in this case to report something is wrong since we could not inspect the status of the ExecCommandAgent, most likely something is wrong.

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go]: Changed to use `retry.RetryNWithBackoffCtx`

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:87]: Removed that check altogether... less code, yay!

File: agent/engine/docker_task_engine.go
Comment (angelcar) [reply agent/engine/docker_task_engine.go:302]: Addresssed

File: agent/engine/docker_task_engine.go
Comment (angelcar) [reply agent/engine/docker_task_engine.go:1260]: So none of the lines below line 1204 will block the start of the container

File: agent/api/container/container.go
Comment (sharanyad) [inline agent/api/container/container.go:273]: this needs a bump in the agent state file version: https://github.com/aws/amazon-ecs-agent/blob/master/agent/statemanager/state_manager.go#L106 (can add a TODO for later too)

File: agent/engine/execcmdagent/manager_linux.go
Comment (sharanyad) [reply agent/engine/execcmdagent/manager_linux.go:40]: we could later think about moving the exec methods in engine to exec agent manager

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:49]: what would happen if it just keeps timing out?

Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go]: Now returning a `RestartStatus` type to indicate if the exec agent was `Restarted`, `NotRestarted`, `Unknown`.
In case of inspect errors, I return `(Unknown, nil)` (i.e. no error)

P.S. I retry the InspectContainerExec command, except for timeout errors, as discussed.

File: agent/api/container/container.go
Comment (angelcar) [reply agent/api/container/container.go:273]: Done. Thanks for pointing this out.

File: agent/engine/docker_task_engine.go
Comment (fierlion) [reply agent/engine/docker_task_engine.go:303]: (thanks for that)
pretty sure this is going to be an enum, and we'll support `PENDING`, `RUNNING`, `STOPPED` states.

File: agent/engine/execcmdagent/manager_linux.go
Comment (fierlion) [inline agent/engine/execcmdagent/manager_linux.go:56]: I think we should use the SSM agent systemd service file as our guide:
`Restart=on-failure`
https://github.com/aws/amazon-ssm-agent/blob/master/packaging/linux/amazon-ssm-agent.service#L10
```
If set to on-failure, the service will be restarted when the process exits with a non-zero exit code, is terminated by a signal (including on core dump, but excluding the aforementioned four signals), when an operation (such as service reload) times out, and when the configured watchdog timeout is triggered.
```

File: agent/engine/execcmdagent/manager.go
Comment (fierlion) [inline agent/engine/execcmdagent/manager.go:37]: We might be in `Unknown` land for longer durations (ex. `dockerTimeoutErr` which you reference in your test).  This makes me ask what we will report when we have `RUNNING`|`PENDING`|`STOPPED` to report.

What does the ECS Agent do for `Unknown` -- are we following that pattern here?
I see the point of it existing, but it brings up a little bit of a tangle.

File: agent/engine/execcmdagent/manager_linux.go
Comment (fierlion) [reply agent/engine/execcmdagent/manager_linux.go:40]: I'm ok with the overhead of these checks -- in case this function gets called from the wrong context we have some simple guard rails.

File: agent/engine/execcmdagent/manager_linux.go
Comment (fierlion) [inline agent/engine/execcmdagent/manager_linux.go:37]: Note: I went back and forth about this function.
I had imagined a bulletproof `start` function that we could call from any context.
I like this approach.

File: agent/engine/execcmdagent/manager_linux.go
Comment (fierlion) [inline agent/engine/execcmdagent/manager_linux.go:99]: debug log here?  why are we calling start on non execute-command enabled tasks?

File: agent/engine/execcmdagent/manager_linux_test.go
Comment (fierlion) [inline agent/engine/execcmdagent/manager_linux_test.go:68]: This test is great.  I love that I can see the lifecycle of the start process directly from these test cases.

File: agent/engine/execcmdagent/manager.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager.go:37]: What would be the alternative? Ignore?

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:56]: So, if the ssm-agent exits with a zero value, we don't want to restart it?

Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:56]: P.S. I don't think we have access to the signals ssm agent receives through the docker API, all we have is the exit code.

File: agent/engine/execcmdagent/manager_linux.go
Comment (angelcar) [reply agent/engine/execcmdagent/manager_linux.go:99]: Will add in next rev

File: agent/engine/execcmdagent/manager_linux.go
Comment (fierlion) [reply agent/engine/execcmdagent/manager_linux.go:57]: Yeah I wonder if we should just always restart n-times, regardless of exit code.  We'll need a good way of surfacing what exit code we're getting.

File: agent/engine/execcmdagent/manager.go
Comment (fierlion) [reply agent/engine/execcmdagent/manager.go:37]: I think agent doesn't update status until it is in a known state.  So we need to find a way of absolutely limiting our time spent in unknown-land.

**Review by fierlion — APPROVED** [review]

**Review by sharanyad — APPROVED** [review]
code lgtm

**Pull request merged**
